Tale workspaces fixes

bodom0015 opened the pull request:
### Problem
Some minor issues remained after the initial work toward supporting Tale Workspaces, namely:
* Fixes #363 
* Fixes #364
* Fixes #365

### Approach
* #363 - Comment out the "Home" button for now
* #364 - Wire up some logic to hit the `/workspace` endpoint instead of `/folder` if we are at the workspace root
* #365 - Renamed "Select Data..." to "Import Tale Data..." (TBD)

### How to Test
1. Check out this branch on your dev machine
2. Rebuild the dashboard code
3. Login to the dashboard
4. Create a few Tales, if you haven't already
5. Navigate to Run > Files > Tale Workspaces
    * NOTE: If you haven't added any data to the Workspace, you won't see any folders/items listed here yet
6. At the top-right of the panel, click the (+) button
    * You should see the dropdown appear beneath the button
    * You should see that the "Select Data..." option has been renamed to "Import Tale Data..." (#365)
7. Click on "Import Tale Data..."
    * The "Select Workspace Data" modal should appear
    * This modal should no longer list "Home" on the left side (#363)
    * You should see that this modal now lists Tale Workspace folders by name instead of ID (#364)
8. Choose some folders / items and click "Move to Workspace" at the bottom-right of the modal
    * The modal should close
    * You should receive notifications in the UI that you files are moving
    * As they are moved, you should see the UI automatically update to display your new Workspace contents
9. At the top-right of the panel, click the (+) button, and choose "Import Tale Data..." again
    * You should see that any folders/items that were successfully moved in the previous step no longer appear in their previous locations, but can now be found by clicking down into your current Tale's subfolder(s)
10. Choose some more folders / items and click "Copy to Workspace" at the bottom-right of the modal
    * The modal should close
    * You should receive notifications in the UI that you files are copying
    * As they are copied, you should see the UI automatically update to display your new Workspace contents

### Notes:
- NOTE: I'm seeing some inconsistencies between "Tale Workspaces" tab and the "Select Workspace Data" modal... I would expect the contents of my Tale's current folder in the modal to match the folders/files in the "Tale Workspaces" tab, but this may not always be the case currently
- NOTE: This submission logic (in `tale-files-tab`) was not modified as a part of this pull request
- NOTE: Calling `/resource/copy` on a tale workspace folder still results in the ID of the tale as the destination folder name - will the EventHandlers need to be updated in `wt_home_dir` as well?
- NOTE: Should we filter out current tale from being moved into itself? this already yields an error popup anyway, but we may need to show this if we want the option to copy a Tale's data into itself (is this useful?)
- NOTE: There is still some failing edge case that I can't reliably recreate - some tales are able to "Move to Workspace" just fine, while others fail with this error popup (see **Verbose Error Details** below). In my case, I was trying to import the workspace data from "new jup" into another tale named "new tale". I was able to move another tale into this workspace already, so I'm not sure why this Tale is presenting a problem.

##### Verbose Error Details
Error popup:
```
{"message": "AssertionError: AssertionError()", "trace": ["<FrameSummary file /girder/girder/api/rest.py, line 630 in endpointDecorator>", "<FrameSummary file /girder/girder/api/rest.py, line 1228 in PUT>", "<FrameSummary file /girder/girder/api/rest.py, line 967 in handleRoute>", "<FrameSummary file /girder/girder/api/access.py, line 63 in wrapped>", "<FrameSummary file /girder/girder/api/describe.py, line 709 in wrapped>", "<FrameSummary file /girder/girder/api/v1/resource.py, line 334 in moveResources>", "<FrameSummary file /girder/girder/models/folder.py, line 338 in move>", "<FrameSummary file /girder/girder/models/model_base.py, line 506 in save>", "<FrameSummary file /girder/girder/events.py, line 314 in trigger>", "<FrameSummary file /girder/plugins/wt_home_dir/server/__init__.py, line 80 in handler>", "<FrameSummary file /girder/plugins/wt_home_dir/server/lib/EventHandlers.py, line 106 in run>", "<FrameSummary file /girder/plugins/wt_home_dir/server/lib/EventHandlers.py, line 133 in moveFolder>", "<FrameSummary file /girder/plugins/wt_home_dir/server/lib/WTFilesystemProvider.py, line 201 in moveRecursive>", "<FrameSummary file /usr/local/lib/python3.5/dist-packages/wsgidav/fs_dav_provider.py, line 313 in moveRecursive>"], "type": "internal"}
```

Failing `tale`:
```
{
    "_accessLevel": 2,
    "_id": "5c476423fb23660001e75c95",
    "_modelType": "tale",
    "authors": "Michael Lambert",
    "category": "science",
    "config": {},
    "created": "2019-01-22T18:42:43.307000+00:00",
    "creatorId": "5c377bd4e9aec300015f72ae",
    "dataSet": [
      {
        "itemId": "5c425c187ccff7000177c90b",
        "mountPath": "Photogrammetric scans of aerial photographs of North American glaciers, 1974.  tiffs - Obliques"
      },
      {
        "itemId": "5c47a139fb236600012d527e",
        "mountPath": "Laboratory incubations investigating controls over the soil priming effect"
      }
    ],
    "description": "asdf",
    "folderId": "5c476423fb23660001e75c97",
    "format": 4,
    "icon": "https://raw.githubusercontent.com/whole-tale/jupyter-base/master/squarelogo-greytext-orangebody-greymoons.png",
    "iframe": true,
    "illustration": "https://raw.githubusercontent.com/whole-tale/dashboard/master/public/images/demo-graph2.jpg",
    "imageId": "5c377a16e9aec300015f71c7",
    "narrative": [],
    "narrativeId": "5c378b9ce9aec300015f7637",
    "public": false,
    "published": false,
    "title": "new jup",
    "updated": "2019-01-24T17:00:57.924000+00:00"
  }
```

Associated `workspace`:
```
{
    "_id": "5c476423fb23660001e75c96",
    "_modelType": "folder",
    "access": {
      "groups": [],
      "users": [
        {
          "flags": [],
          "id": "5c377bd4e9aec300015f72ae",
          "level": 2
        }
      ]
    },
    "baseParentId": "5c378b9ce9aec300015f762f",
    "baseParentType": "collection",
    "created": "2019-01-22T18:42:43.309000+00:00",
    "creatorId": null,
    "description": "",
    "lowerName": "new jup",
    "meta": {
      "taleId": "5c476423fb23660001e75c95"
    },
    "name": "new jup",
    "parentCollection": "folder",
    "parentId": "5c378b9ce9aec300015f7630",
    "public": false,
    "size": 0,
    "updated": "2019-01-22T18:42:43.312000+00:00"
  }
```

Xarthisius commented:
This looks great! I have a couple remarks about UX:
1. When I click "move to workspace", the item/folder that I'm moving is not disappearing right away. I have to navigate up/down one level and go back to refresh the view. Move operation should probably refresh the view.
1. Surprisingly, I can hit 'move' multiple times and every time operation finishes with a success (according to the toast that appear in the background). Should be fixed by the former. 
1. I'd argue that on a "workspace root level" buttons shouldn't be active, i.e. there's shouldn't be a way for a user to copy/move entire workspace as a folder into the current one. The latter would be particularly disastrous.

craig-willis commented:
Overall, looking good.  I'm also seeing @Xarthisius' issue 2 and agree with 4.   Per your test case, the modal should close on move/copy, which isn't happening for me.  

Also, I'm not certain about the "Import Tale Data..." label change.  Eventually, I think we'll want to be able to move/copy from home (when fixed), but maybe I missed something.

bodom0015 commented:
@Xarthisius 
1. I cannot reliably recreate this issue, nor can I reliably recreate @craig-willis' issue where the modal was refusing to close... more details/discussion may be needed
2. Since the modal closes and the view updates properly for me, I cannot seem to pin this one down either
3. Good call, I've added some logic to disable the button when the user is at the root level
4. I had added logic previously to filter, but since Tale names are not unique we actually needed to filter by `taleId`.. I've added some dumb logic to do this 

@craig-willis 
> Also, I'm not certain about the "Import Tale Data..." label change. Eventually, I think we'll want to be able to move/copy from home (when fixed), but maybe I missed something.

I am open to alternatives here... maybe just "Import Data into Tale" or something similar?

bodom0015 commented:
Added a bit to explicitly close the modal after a successful move or copy operation - hopefully this fixes the sync issues @Xarthisius is seeing as well, but if not I can add a call to the "resync" handler as well.

Until this and the PR with the view changes for the `/dataset` endpoint are merged, there may be synchronization issues that are fixed in one branch and not another.

Xarthisius commented:
Modal now closes for me on each operation, however I do get this in the console:

```
component.js:518 Uncaught TypeError: Cannot read property 'then' of undefined
    at Class.updateWorkspaceData (component.js:518)
    at Backburner._join (backburner.js:741)
    at Backburner.join (backburner.js:477)
    at Function.run.join (ember-metal.js:456)
    at ember-glimmer.js:974
    at exports.flaggedInstrument (ember-metal.js:4582)
    at Class.updateWorkspaceData (ember-glimmer.js:973)
    at Class.updateWorkspaceData (component.js:70)
    at Backburner._run (backburner.js:758)
    at Backburner._join (backburner.js:736)
```

I'm gonna try with a clean env and report back.

bodom0015 commented:
Fixed that error in the console, and made it so that the filtering on `meta.taleId` is only done at the workspace root level

craig-willis commented:
LGTM. WFM

bodom0015 commented on app/models/workspace.js:
```diff
@@ -0,0 +1,33 @@
+  _modelType: DS.attr('string'),
+  baseParentType: DS.attr('string'), // folder or item
+  baseParentId: DS.attr('string'), // folder or item
+  // baseParentId: Ember.computed('baseParentType', function() {
```
This class was copied from the existing model for `folder`.. I was unsure what the commented out fields on that model were for, but preserved them here in case they were desirable.

I'd be happy to replace them with a shorter comment referencing the original implementation
